Feat/review todo section

joosomi opened the pull request:
### **PR Type**
Enhancement, Documentation


___

### **Description**
- Add support for scanning and reporting TODO comments in PR code.
  - Introduce `require_todo_scan` config and prompt options.
  - Render TODO sections in markdown review output.
  - Define new `TodoSection` schema for structured TODO reporting.

- Update documentation to describe the new TODO scan feature.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Render TODO comment sections in markdown review output</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/algo/utils.py

<li>Add markdown rendering for TODO sections in review output.<br> <li> Display TODOs with file, line, and content, supporting GFM and plain <br>markdown.<br> <li> Integrate TODO section with emoji and conditional formatting.


</details>


  </td>
  <td><a href="https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53">+45/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pr_reviewer.py</strong><dd><code>Add require_todo_scan to reviewer prompt context</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_reviewer.py

- Pass `require_todo_scan` setting to reviewer prompt context.


</details>


  </td>
  <td><a href="https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-8e265068e189a06852605cc694b03e92b523b4f8162077a2f3455ced4cdad8dc">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pr_reviewer_prompts.toml</strong><dd><code>Add TodoSection schema and todo_sections field to prompts</code></dd></summary>
<hr>

pr_agent/settings/pr_reviewer_prompts.toml

<li>Define <code>TodoSection</code> schema for TODO comment reporting.<br> <li> Add <code>todo_sections</code> field to Review model, conditionally included.<br> <li> Update example outputs to show TODO section usage.


</details>


  </td>
  <td><a href="https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-2e1ca45d2d8635b5f9cde801d9d210f6516e7f15d45dd9b0be134ac91e7a2e63">+12/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>review.md</strong><dd><code>Document require_todo_scan configuration option</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/docs/tools/review.md

<li>Document <code>require_todo_scan</code> configuration option.<br> <li> Explain its effect on PR review output.


</details>


  </td>
  <td><a href="https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-8e6363749b8b0a82468e5fbb0796cee806dc48e70e7472b04088d56d2c415094">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>configuration.toml</strong><dd><code>Enable require_todo_scan in reviewer configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/settings/configuration.toml

- Add `require_todo_scan=true` to default reviewer settings.


</details>


  </td>
  <td><a href="https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-66cfda5143e484ee53ecf7aa0df7dca8ad0b181256f4b0675905db35bcbbae78">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

joosomi committed:
feat: add TODO comments to PR review output

joosomi committed:
fix: avoid incorrect ToDo header

Judy-Choi commented on pr_agent/algo/utils.py:
```diff
@@ -209,6 +210,50 @@ def convert_to_markdown_v2(output_data: dict,
+            if gfm_supported:
+                markdown_text += f"<tr><td>"
+                if is_value_no(value):
+                    markdown_text += f"{emoji}&nbsp;<strong>No ToDo sections</strong>"
```
markdown_text += f"{emoji}&nbsp;<strong>No TODO sections</strong>"

Judy-Choi commented on pr_agent/settings/pr_reviewer_prompts.toml:
```diff
@@ -102,6 +107,9 @@ class Review(BaseModel):
     security_concerns: str = Field(description="Does this PR code introduce possible vulnerabilities such as exposure of sensitive information (e.g., API keys, secrets, passwords), or security concerns like SQL injection, XSS, CSRF, and others ? Answer 'No' (without explaining why) if there are no possible issues. If there are security concerns or issues, start your answer with a short header, su ...
 {%- endif %}
+{%- if require_todo_scan %}
+    todo_sections: Union[List[TodoSection], str] = Field(description="A list of TODO comments found in the code. Return 'No' (as a string) if there are no TODO comments.")
```
todo_sections: List[TodoSection] = Field(description="A list of TODO comments found in the code. Return an empty list if there are no TODO comments.")

TODO section이 없는 경우 빈 리스트 `[]` 을 리턴합니다.

Judy-Choi replied:
단, 이 프롬프트만 수정할 경우 TODO 가 없는 경우 최종 review 에 No TODO 섹션이 생성되지 않습니다.
<img width="911" alt="image" src="https://github.com/user-attachments/assets/08298565-53c6-49da-b760-6de10100d52c" />

No TODO 섹션이 생성되지 않는 이유는 utils.py 에서 빈 리스트 `[]` 가 걸러지기 때문.
```python
    for key, value in output_data['review'].items():
        if value is None or value == '' or value == {} or value == []:
```

따라서 No TODO 섹션이 생성되도록 하려면
1. 기존 코드를 유지하여 TODO section이 없는 경우 'No' 를 리턴하거나
2. utils.py 의 조건문을 수정하여 `todo_sections=[]` 인 경우 섹션을 생성해야 함.

개인적인 의견으로 저는 기존 코드와의 통일성을 고려하여 기존 코드를 유지하는 편이 좋을 것 같습니다.

Judy-Choi commented on pr_agent/algo/utils.py:
```diff
@@ -209,6 +210,50 @@ def convert_to_markdown_v2(output_data: dict,
+                if is_value_no(value):
+                    markdown_text += f"{emoji}&nbsp;<strong>No ToDo sections</strong>"
+                else:
+                    markdown_text += f"{emoji}&nbsp;<strong>ToDo sections</strong><br><br>\n\n"
```
TODO

kir-rin committed:
fix: remove whitespace from relevant file

kir-rin commented on pr_agent/algo/utils.py:
```diff
@@ -220,7 +220,7 @@ def convert_to_markdown_v2(output_data: dict,
                         markdown_text += "<ul>\n"
                         for todo_item in value:
-                            relevant_file = todo_item.get('relevant_file', '')
+                            relevant_file = todo_item.get('relevant_file', '').strip()
```
Gitlab에서는 relevant_file에 공백 문자 \n이 들어가면 링크가 깨져서 수정했습니다! https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/b83085ea00bef0fd6052365156141dead226fcc7
<img width="711" alt="image" src="https://github.com/user-attachments/assets/ade85fe7-0c79-465f-8cad-a82901a2b3fa" />
<img width="821" alt="image" src="https://github.com/user-attachments/assets/9c674175-0fb9-41b2-bf90-1d46db480bbb" />

kir-rin reviewed (approved):
문서 수정까지 감사합니다. 수고하셨습니다! LGTM 👍

joosomi commented:
## Generating PR code suggestions


Work in progress ...<br>
<img src="https://codium.ai/images/pr_agent/dual_ball_loading-crop.gif" width=48>

joosomi committed:
refactor: improve todo section handling and clarify todo content description

joosomi commented:
## Generating PR code suggestions


Work in progress ...<br>
<img src="https://codium.ai/images/pr_agent/dual_ball_loading-crop.gif" width=48>

benedict-lee commented on pr_agent/algo/utils.py:
```diff
@@ -209,6 +210,51 @@ def convert_to_markdown_v2(output_data: dict,
+                    if git_provider and relevant_file and line_number:
+                        reference_link = git_provider.get_line_link(relevant_file, line_number, line_number)
+                except Exception as e:
+                    print(f"Error generating link: {e}")
```
print로 출력, 로그 메시지

Judy-Choi replied:
로그 메시지 출력으로 수정했습니다.
[refactor: Change error messsage 'print' -> 'get_logger'](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/7d38814cae372d530cd1ed8f39ef8ae3abbe28df)
[7d38814](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/7d38814cae372d530cd1ed8f39ef8ae3abbe28df)

```python
get_logger().exception(f"Error generating link: {e}")
```

benedict-lee commented on pr_agent/algo/utils.py:
```diff
@@ -209,6 +210,51 @@ def convert_to_markdown_v2(output_data: dict,
+                    print(f"Error generating link: {e}")
+                    return f"{relevant_file} [{line_number}]: {content}"
+
+                file_line = f"{relevant_file} [{line_number}]"
```
변수명

Judy-Choi replied:
@benedict-lee  멘토님 질문 있습니다 🙋🏻‍♀️
해당 변수에는 `파일명[TODO 주석이 있는 라인 번호]` 값이 저장되는데요,
`file_line` 이라는 변수명이 직관적이지 않아서 수정이 필요하다는 뜻인가요?
(ChatGPT 는 `file_location` 이라는 이름을 추천해 주었습니다 🤓)

아니면 `file_line` 변수가 쓰이는 전체 코드는 다음과 같은데, 리팩토링이 필요하다는 뜻인가요?
```python
                file_line = f"{relevant_file} [{line_number}]"
                if reference_link:
                    if gfm_supported:
                        file_line = f"<a href='{reference_link}'>{file_line}</a>"
                    else:
                        file_line = f"[{file_line}]({reference_link})"
                return f"{file_line}: {content}" if content.strip() else file_line
```

benedict-lee commented on pr_agent/algo/utils.py:
```diff
@@ -209,6 +210,51 @@ def convert_to_markdown_v2(output_data: dict,
                     value = emphasize_header(value.strip(), only_markdown=True)
                     markdown_text += f"{value}\n\n"
+        elif 'todo sections' in key_nice.lower():
+            def format_todo_item(todo_item):
```
헬퍼이긴하지만 타입힌트

Judy-Choi replied:
타입 힌트 추가했습니다.
[refactor: add TypedDict and type hints to todo item formatter](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/53b913a4cb6c16f8a9fee3a9489c4935adbef1cb)
[53b913a](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/53b913a4cb6c16f8a9fee3a9489c4935adbef1cb)

benedict-lee commented on pr_agent/settings/pr_reviewer_prompts.toml:
```diff
@@ -72,6 +72,11 @@ class KeyIssuesComponentLink(BaseModel):
 
+class TodoSection(BaseModel):
+    relevant_file: str = Field(description="The file containing the TODO comment")
+    line_number: int = Field(description="The line number of the TODO comment")
```
싱글라인만 지원인가요?

Judy-Choi replied:
TODO 가 싱글라인일 경우에 대해서만 개발하였으나, 멘토님께서 지적해 주셔서 멀티라인도 지원하도록 기능 추가했습니다.
([feat: TODO multi-line](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/8c7c087931154fc4d6d7e5b40553b9c961271d1d))
최종 PR 시 멀티라인 지원 여부를 결정해야 하고, 만약 지원할 경우 blockquote 를 사용할지 여부도 함께 결정해야 할 것 같습니다.
따라서 다음 3가지 중 1가지를 결정하여 최종 PR 하는 방안을 제안드리고 싶습니다.

### 제안방법
#### 1. 싱글라인 표기만 지원
#### 2. 멀티라인 지원
- 아래 캡쳐에서 [12-14] 에 해당
#### 3. 멀티라인 지원 - block으로 묶기
- 아래 캡쳐에서 [18-20] 에 해당

<img width="438" alt="image" src="https://github.com/user-attachments/assets/ba52368b-f314-483b-90c7-5a3aa2790ccc" />

Judy-Choi commented:
+ 멘토님 피드백 : TODO 가 많을 경우 섹션이 접히면 좋겠습니다

Judy-Choi committed:
modify: ToDo -> TODO

Judy-Choi commented:
https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/bab8ee96334b87e8c6d4bd65b0c552bf200bd2f0 커밋에서
`ToDo` -> `TODO` 로 일괄 수정하였습니다.

Judy-Choi committed:
feat: Add <details> to TODO sections

Judy-Choi commented:
> * 멘토님 피드백 : TODO 가 많을 경우 섹션이 접히면 좋겠습니다

피드백 반영 완료 : [feat: Add `<details>` to TODO sections](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/08440d8ebd3e1ca2283f30023a5ff93f159514d3) 커밋

## 예시
### TODO 가 있는 경우
<img width="919" alt="image" src="https://github.com/user-attachments/assets/c8d5e32e-ce44-4376-8f18-47746585b1e5" />
<img width="919" alt="image" src="https://github.com/user-attachments/assets/9bb84327-ba35-4318-9dc2-14b614094194" />

### TODO 가 없는 경우 (No TODO)
접기 화살표 ► 가 표기되지 않음
<img width="919" alt="image" src="https://github.com/user-attachments/assets/b103ee8f-fdff-4e49-a01c-d79cab35ef42" />

Judy-Choi committed:
refactor: Change error messsage 'print' -> 'get_logger'

Judy-Choi committed:
refactor: add TypedDict and type hints to todo item formatter

Judy-Choi committed:
feat: TODO multi-line

kir-rin committed:
feat: include item count in TODO sections summary

kir-rin committed:
feat: add todos_summary to review output in markdown conversion

kir-rin committed:
refactor: remove unused HTML formatting function from utils

kir-rin committed:
Merge pull request #3 from PullPullers/feedback/review-todo-section

Feedback/review todo section

joosomi commented:
## Generating PR code suggestions


Work in progress ...<br>
<img src="https://codium.ai/images/pr_agent/dual_ball_loading-crop.gif" width=48>

kir-rin committed:
refactor: rename file line to file reference

joosomi commented:
## PR Code Suggestions ✨
<!-- dc9e9af -->

Latest suggestions up to dc9e9af

<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>General</td>
<td>



<details><summary>Improve error handling in TODO formatting<!-- not_implemented --></summary>

___


**The <code>format_todo_item</code> function lacks error handling for cases where <code>line_range</code> is not <br>a tuple or list.  Add a check to handle this scenario gracefully, perhaps by logging <br>a warning and returning a default string.  Consider using more descriptive variable <br>names for better readability.**

[pr_agent/algo/utils.py [227-261]](https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53R227-R261)

```diff
 def format_todo_item(todo_item: TodoItem) -> str:
-    relevant_file = todo_item.get('relevant_file', '').strip()
-    line_range = todo_item.get('line_range', [])
+    file_path = todo_item.get('relevant_file', '').strip()
+    line_range_str = todo_item.get('line_range', [])
     content = todo_item.get('content', '')
     reference_link = None
-    if isinstance(line_range, str):
-        line_range = ast.literal_eval(line_range.strip())
+
     try:
-        if git_provider and relevant_file and line_range:
-            reference_link = git_provider.get_line_link(relevant_file, line_range[0], line_range[1])
-    except Exception as e:
-        get_logger().exception(f"Error generating link: {e}")
-        line_str = f"[{line_range[0]}]" if line_range[0] == line_range[1] else f"[{line_range[0]}-{line_range[1]}]"
-        return f"{relevant_file} {line_str}: {content}"
+        if isinstance(line_range_str, str):
+            line_range = ast.literal_eval(line_range_str.strip())
+        elif isinstance(line_range_str, (tuple, list)) and len(line_range_str) == 2:
+            line_range = tuple(line_range_str)
+        else:
+            get_logger().warning(f"Invalid line_range format: {line_range_str} for file: {file_path}. Using default.")
+            return f"{file_path}: Invalid line range"
+
+        if git_provider and file_path and line_range:
+            reference_link = git_provider.get_line_link(file_path, line_range[0], line_range[1])
+    except (ValueError, SyntaxError, TypeError) as e:
+        get_logger().exception(f"Error processing line range: {e} for file: {file_path}. Using default.")
+        return f"{file_path}: Error processing line range"
+
     line_str = f"[{line_range[0]}]" if line_range[0] == line_range[1] else f"[{line_range[0]}-{line_range[1]}]"
-    file_ref = f"{relevant_file} {line_str}"
+    file_ref = f"{file_path} {line_str}"
     if reference_link:
         if gfm_supported:
             file_ref = f"<a href='{reference_link}'>{file_ref}</a>"
         else:
             file_ref = f"[{file_ref}]({reference_link})"
     content_lines = content.strip().split("\n")
     if len(content_lines) == 1:
         return f"{file_ref}: {content_lines[0]}"
     elif len(content_lines) > 1:
         content_lines = "<br>".join(content_lines)
         return f"{file_ref}: <blockquote>\n{content_lines}\n</blockquote>"
     else:
         return file_ref
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: The suggestion improves the robustness of the `format_todo_item` function by adding error handling for invalid `line_range` formats.  This prevents unexpected crashes and improves the overall reliability of the code.  The score is not higher because it's primarily an improvement in error handling, not a fix for a critical bug.


</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Simplify conditional value checks</summary>

___


**The condition <code>value is None or value == '' or value == {} or value == []</code> can be <br>simplified.  Checking for truthiness (<code>if not value:</code>) will cover all these cases <br>except for empty dictionaries, which should be handled separately if needed.  This <br>improves readability and reduces redundancy.**

[pr_agent/algo/utils.py [169-171]](https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53R169-R171)

```diff
-if value is None or value == '' or value == {} or value == []:
-            if key.lower() not in ['can_be_split', 'key_issues_to_review']:
+if not value and key.lower() not in ['can_be_split', 'key_issues_to_review']:
                 continue
             if key.lower() == 'can_be_split':
                 emoji = "🔀"
```
<details><summary>Suggestion importance[1-10]: 5</summary>

__

Why: The suggestion simplifies the conditional check, improving readability. However, the impact is relatively minor, as the original code was already functional.  The change doesn't address a significant issue or improve performance.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

___

#### Previous suggestions


<details><summary>Suggestions up to commit dc64603</summary>
<br><table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=1>General</td>
<td>



<details><summary>Simplify conditional formatting logic<!-- not_implemented --></summary>

___


**The <code>format_todo_item</code> function uses a nested <code>if/elif/else</code> structure to handle <br>different content lengths. This can be simplified using a more concise approach.  <br>Consider using a single conditional expression to determine the appropriate <br>formatting based on the length of <code>content_lines</code>.**

[pr_agent/algo/utils.py [227-261]](https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53R227-R261)

```diff
 def format_todo_item(todo_item: TodoItem) -> str:
     relevant_file = todo_item.get('relevant_file', '').strip()
     line_range = todo_item.get('line_range', [])
     content = todo_item.get('content', '')
     reference_link = None
     if isinstance(line_range, str):
         line_range = ast.literal_eval(line_range.strip())
     try:
         if git_provider and relevant_file and line_range:
             reference_link = git_provider.get_line_link(relevant_file, line_range[0], line_range[1])
     except Exception as e:
         get_logger().exception(f"Error generating link: {e}")
         line_str = f"[{line_range[0]}]" if line_range[0] == line_range[1] else f"[{line_range[0]}-{line_range[1]}]"
         return f"{relevant_file} {line_str}: {content}"
     line_str = f"[{line_range[0]}]" if line_range[0] == line_range[1] else f"[{line_range[0]}-{line_range[1]}]"
     file_line = f"{relevant_file} {line_str}"
     if reference_link:
-        if gfm_supported:
-            file_line = f"<a href='{reference_link}'>{file_line}</a>"
-        else:
-            file_line = f"[{file_line}]({reference_link})"
-    content_lines = content.strip().split("\n")
-    # if TODO content is single-line :
-    if len(content_lines) == 1:
-        return f"{file_line}: {content_lines[0]}"
-    # else if TODO content is multi-line:
-    elif len(content_lines) > 1:
-        content_lines = "<br>".join(content_lines)
-        return f"{file_line}: <blockquote>\n{content_lines}\n</blockquote>"
-    # else if TODO content is empty:
-    else:
-        return file_line
+        file_line = f"<a href='{reference_link}'>{file_line}</a>" if gfm_supported else f"[{file_line}]({reference_link})"
+    content_lines = content.strip().split('\n')
+    return f"{file_line}: {content_lines[0]}" if len(content_lines) == 1 else f"{file_line}: <blockquote>{'<br>'.join(content_lines)}</blockquote>" if len(content_lines) > 1 else file_line
```
<details><summary>Suggestion importance[1-10]: 5</summary>

__

Why: The suggestion improves the readability of the `format_todo_item` function by replacing a nested `if/elif/else` structure with a more concise conditional expression.  The impact is moderate, improving code clarity but not significantly altering functionality.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

</details>
<details><summary>✅ Suggestions up to commit ca3df35</summary>
<br><table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>General</td>
<td>



<details><summary>✅ <s>Improve exception handling in link generation</s></summary>

___

<details><summary><b>Suggestion Impact:</b></summary>The commit improved error handling by replacing the print statement with proper logging using get_logger().exception(), which provides better error context and debugging capabilities

code diff:

```diff
                 except Exception as e:
-                    print(f"Error generating link: {e}")
+                    get_logger().exception(f"Error generating link: {e}")
```

</details>


___


**The <code>try...except</code> block catches general exceptions during link generation. This is <br>too broad and might mask unexpected errors.  Consider catching more specific <br>exceptions (e.g., <code>AttributeError</code>, <code>TypeError</code>) or logging the error with more context <br>instead of just printing it to the console.  This will improve debugging and error <br>handling.**

[pr_agent/algo/utils.py [219-224]](https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53R219-R224)

```diff
 try:
     if git_provider and relevant_file and line_number:
         reference_link = git_provider.get_line_link(relevant_file, line_number, line_number)
-except Exception as e:
-    print(f"Error generating link: {e}")
+except (AttributeError, TypeError) as e:
+    import logging
+    logging.error(f"Error generating link for {relevant_file}:{line_number}: {e}")
     return f"{relevant_file} [{line_number}]: {content}"
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: The suggestion improves error handling by catching more specific exceptions and logging the error with more context, which aids in debugging and identifying the root cause of link generation failures.  This is a moderate improvement in robustness.


</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Improve non-GFM TODO list formatting<!-- not_implemented --></summary>

___


**The code handles both GFM and non-GFM formats for displaying TODO sections.  <br>However, the non-GFM formatting uses a simple bullet list (<code>-</code>), which might not be <br>visually distinct enough. Consider using a more structured format like a numbered <br>list (<code>1.</code>, <code>2.</code>, etc.) for better readability in non-GFM environments.  This will <br>improve the clarity and organization of the TODO list when GFM is not supported.**

[pr_agent/algo/utils.py [234-257]](https://github.com/PullPullers/pr-agent-qodo/pull/1/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53R234-R257)

```diff
 if gfm_supported:
     markdown_text += f"<tr><td>"
     if is_value_no(value):
         markdown_text += f"{emoji}&nbsp;<strong>No ToDo sections</strong>"
     else:
         markdown_text += f"{emoji}&nbsp;<strong>ToDo sections</strong><br><br>\n\n"
         if isinstance(value, list):
             markdown_text += "<ul>\n"
-            for todo_item in value:
-                markdown_text += f"<li>{format_todo_item(todo_item)}</li>\n"
+            for i, todo_item in enumerate(value):
+                markdown_text += f"<li>{i+1}. {format_todo_item(todo_item)}</li>\n"
             markdown_text += "</ul>\n"
         else:
             markdown_text += f"<p>{format_todo_item(value)}</p>\n"
     markdown_text += f"</td></tr>\n"
 else:
     if is_value_no(value):
         markdown_text += f"### {emoji} No ToDo sections\n\n"
     else:
         markdown_text += f"### {emoji} ToDo sections\n\n"
         if isinstance(value, list):
-            for todo_item in value:
-                markdown_text += f"- {format_todo_item(todo_item)}\n"
+            for i, todo_item in enumerate(value):
+                markdown_text += f"{i+1}. {format_todo_item(todo_item)}\n"
         else:
             markdown_text += f"- {format_todo_item(value)}\n"
```
<details><summary>Suggestion importance[1-10]: 5</summary>

__

Why: The suggestion improves the readability of the TODO list in non-GFM environments by using a numbered list instead of a simple bullet list.  However, the impact is relatively minor, as it only affects the visual presentation.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

</details>

kir-rin committed:
feat: toggle open or closed based on line count for todo section

kir-rin committed:
refactor: standardizes todo_sections and todos_summary formatting

kir-rin committed:
fix: use | for multiline in YAML instead of \n for todo content prompts

kir-rin committed:
refactor: handle singular/plural forms of entry

kir-rin committed:
refactor: remove count info from todo summary

Judy-Choi committed:
Merge pull request #4 from PullPullers/feat/toggle-open-based-on-line-count-for-todo-section

todo section의 내용이 10줄 이하이면 기본적으로 펼쳐져 있도록 수정

Judy-Choi committed:
fix: display no TODO comments section

Judy-Choi reviewed (commented):
[fix: display no TODO comments section](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/ca05b798caae156168a7cdd4da81486116eafc49)
[ca05b79](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/ca05b798caae156168a7cdd4da81486116eafc49)

TODO 주석이 없을 때 LLM의 output 이 일관되지 않게 생성됩니다.
ex) `todo_sections` = [], False (원래 프롬프트대로라면 'No' 가 들어가야 함.

우선 다음과 같이 해결했습니다.
1. 프롬프트 수정
2. TODO 주석이 없는 경우 즉시 No TODO sections 생성
markdown_todo_items = format_todo_items(value) 즉 value 로부터 TODO 주석 정보를 가져오는 코드의 위치를 value 값을 체크하는 코드 다음에 위치시켜 value 가 False 나 'No' 인 경우에 즉시 No TODO sections 를 생성하도록 수정했습니다.

Judy-Choi committed:
feat: Change default TODO flag value from true to false

Judy-Choi committed:
fix: update default value of require_todo_scan from true to false in docs

Judy-Choi committed:
style: replace todo emoji with 📝

Judy-Choi committed:
style: refine layout of TODO section (add <br><br>)

Judy-Choi committed:
feat: improve TODO sections closed by default

Judy-Choi commented:
TODO 섹션이 항상 기본적으로 접혀 있도록 수정 후 테스트 결과 첨부
[feat: improve TODO sections closed by default](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/8325a8aeb1c9d31b58d8c5f05e7bb66af4b1774d)
[8325a8a](https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/8325a8aeb1c9d31b58d8c5f05e7bb66af4b1774d)

**TODO가 10lines 이하**
기존에는 섹션이 펼쳐져 보였으나, 접혀서 보이도록 수정
https://github.com/Judy-Choi/pr-agent-study/pull/11
![image](https://github.com/user-attachments/assets/e221edbf-0d6a-4cf9-ab41-2d6dcb6c0d9f)

**TODO 가 10lines 이상**
기존대로 섹션이 접혀서 보임
https://github.com/Judy-Choi/pr-agent-study/pull/8
![image](https://github.com/user-attachments/assets/66c1c1c1-6eca-428d-81bd-93947a766ef9)

**TODO 가 없는 경우**
기존 테스트 url 사용할 경우 FIXME 가 추출되는 버그가 있어 url 교체하여 테스트.
FIXME 추출 버그는 추후 수정 예정.
https://github.com/Judy-Choi/pr-agent-study/pull/1
![image](https://github.com/user-attachments/assets/6e0b8dc6-0e46-4229-925b-8609cf0a053c)

kir-rin committed:
style: remove entry info from the todo section header

related comment: https://github.com/qodo-ai/pr-agent/pull/1869#issuecomment-2979088345

kir-rin committed:
fix: shorten todo summary prompt for balance

related comment: https://github.com/qodo-ai/pr-agent/pull/1869#discussion_r2151443243

kir-rin commented:
Todo Summary 단축 (https://github.com/PullPullers/pr-agent-qodo/pull/1/commits/3d34d4b8432fe3d29ead9b18fe8d7c9fe768289e) 테스트 결과 첨부

이전
![image](https://github.com/user-attachments/assets/fd65c665-ffaf-4b68-98d8-4123f2a46bf6)

이후
![image](https://github.com/user-attachments/assets/08fcd47c-3dec-437a-8f2d-dea4bfaf1f07)

TODO가 없을 때 No라고 잘 뜨는 모습 
(LLM의 output에 todo_summary가 No로 나오지 않아도 상관 없음)
![image](https://github.com/user-attachments/assets/cce70af7-6878-4f5e-8fce-fece7477704e)

Judy-Choi committed:
refactor: clarify TODO field description and simplify 'No' return

kir-rin committed:
fix: exclude TODO prompt if require_todo_scan is false

kir-rin commented on pr_agent/settings/pr_reviewer_prompts.toml:
```diff
@@ -72,10 +72,12 @@ class KeyIssuesComponentLink(BaseModel):
     start_line: int = Field(description="The start line that corresponds to this issue in the relevant file")
     end_line: int = Field(description="The end line that corresponds to this issue in the relevant file")
 
+{%- if require_todo_scan %}
 class TodoSection(BaseModel):
     relevant_file: str = Field(description="The file containing the TODO comment")
     line_range: Tuple[int, int] = Field(description="Start and end line numbers of the TODO comment (inclusive). Must be a tuple of two integers, e.g., (7, 7) for a single line or (7, 10) for a range. Do not use list format [7, 7].")
     content: str = Field(description="The content of the TODO comment. Only include actual TODO comments within code comments (e.g., lines starting with '#', '//', '/*', '<!--').  Remove leading 'TODO' prefixes. Include lines like '# TODO', even if empty after prefix removal. Exclude TODOs outside comments or without appropriate prefixes. Use | block style only when the line range covers multiple ...
+{%- endif %}
```
<img width="296" alt="image" src="https://github.com/user-attachments/assets/d6e15f90-76c8-43af-9c14-061cf3f0ffc4" />

require_todo_scan이 false여도 섹션이 추가되는 버그가 있어
다른 설정의 위 프롬프트 참고해서 config가 true일 때만 포함되도록 수정했습니다!

Judy-Choi replied:
테스트 결과 `require_todo_scan=false` 인 경우에 TODO 섹션 생성되지 않습니다. 감사합니다 윤희님!
